### ISSUE TITLE
Optimized kmk built with GCC 9 fails with error

### ISSUE THREAD

**Issue by dmik** [bitwiseworks/kbuild-os2#3]
An attempt to start `kmk` from RPM version 1:0.1.9998-13 and later immediately ends up with the following error:
```
kmk.exe: *** virtual memory exhausted.  Stop.
```
Debugging shows that the guilty function is `variable_expand_string_2` in `src/kmk/expand.c`. In this fragment:
```c
  /* Simple 2: Nothing to expand. ~50% if the kBuild calls. */

  p1 = (const char *)memchr (string, '$', length);
  if (p1 == 0)
    {
```
the condition `p1 == 0` is never true even when `memchr` returns a null pointer.

Such a behavior is caused by the GCC optimizer. RPM builds of kbuild use its own optimize settings which are -O3. If I downgrade it to -O0, the problem goes away. All other optimize levels don't fix it. It's really strange, smells like a GCC optimizer bug to me.

**Comment by dmik** [comment]
I was able to reduce it to a very simple test case that fails too:
```c
#include <stdio.h>
#include <string.h>

#undef memchr
#define memchr(a,b,c) __my_memchr((a),(b),(c))
static inline void *__my_memchr (__const void *__s, int __c, size_t __n)
{
  register unsigned long int __d0;
  register unsigned char *__res;
  if (__n == 0)
    return NULL;
  __asm__ __volatile__
    ("repne; scasb\n\t"
     "je	1f\n\t"
     "movl	$1,%0\n"
     "1:"
     : "=D" (__res), "=&c" (__d0)
     : "a" (__c), "0" (__s), "1" (__n),
       "m" ( *(struct { __extension__ char __x[__n]; } *)__s)
     : "cc");
  return __res - 1;
}

int main(int argc, const char **argv)
{
  const char *cp = memchr("abc", 'd', 3);
  int eq = cp == 0;
  printf("cp %p eq %d\n", cp, eq);

  return 0;
}
```

This `memchr` definition is from `src/kmk/make.h`. Some sort of optimization which is active only on EMX (OS/2) and only in release builds (debug builds use the normal GCC memchr builtin which works fine). If this test is compiled with -O1..-O3, eq is always fail. Commenting out the memchr override also fixes the problem — the test starts working fine with any optimization level.

Now I need to understand why this cryptic assembly gives some strange results.

**Comment by dmik** [comment]
I've fixed it in the commit above. Apparently, the GCC 9 optimizer is not happy about __res being modified outside the assembly (or doesn't expect it is set to 1 inside the assembly) and somehow assumes that its **never** 0 so any comparison of the return value with 0 is simply optimized out to be always zero (I checked the generated assembly).

I don't know if it's a GCC optimizer bug or the original assembly was not correct and I'm not going to sort that out within this ticket. I can only say that I checked GCC 9.2.1 on Ubuntu and it shows exactly the same failure when compiled with `-m32 -O3` (i.e. in 32 bit mode; in 64 bit mode the failure does not show up regardless of the optimization level).
